### DISCUSSION TITLE
Trying to update Interfaces via API but get error.

### DISCUSSION THREAD

**Discussion by NordloSydost-Pajje** [netbox-community/netbox#12821] [Help Wanted!] [answered]
Hi everyone,

I'm trying to update interfaces for devices via API from PHP, When i try to update the mac-address i get an error from the API:

400 Bad Request` response: {"mac_address":["Invalid MAC address format: unknown"]}

I'm using this url: https://xxxxxxxxx/api/dcim/interfaces/1725/ and trying to post the following Json:

{
  "device": "58",
  "name": "Bridge-Aggregation1",
  "type": "virtual",
  "enabled": true,
  "label": "Some Text",
  "speed": "20000000",
  "mac_address": "XXXXXXXXXXXX"
}

I have tried the mac-address in several ways, with and without :, but are still getting an error. 

When i try the same json via swagger it works (getting response 200).

I dont know what i'm doing wrong!

Br
Per-Åke

**Comment by candlerb** [comment]
> When i try the same json via swagger it works (getting response 200).

It sounds like a problem with your PHP code then - can you post the code which reproduces the problem?

You could also try tcpdump on the Netbox server (`tcpdump -i lo -nn -s0 -A tcp port 8001`) to inspect the HTTP request body from Swagger and the HTTP request body from PHP, and see how they differ.  I would guess that the mac_address field isn't being encoded properly by PHP.

**Comment by NordloSydost-Pajje** [comment]
This is the PHP-code that makes the call:

        $options = [
            "device" => $device_id,
            "name" => (string) $interfacename,
            //"type" => (string) 'virtual',
            "enabled" =>(bool) $interface['is_enabled'],
            "label" => (string) $interface['description'],
            "speed" => (string) $interface['speed']
        ];
        if ($interface['mtu'] > 0)
        {
            $options += [ "mtu" => (string) $interface['mtu'] ];
        }
        if(!$interface['mac_address'] != 'unknown')
        {
            $options += [ "mac_address" => (string) str_replace(":", "", $interface['mac_address']) ]; //This is me trying a different way
        }
        //dd(str_replace(":", "", $interface['mac_address']));
        $url = $this->apihost . '/api/dcim/interfaces/' . $interfaceid . '/';
        dd(json_encode($options), $url);
        $response = Http
            ::withHeaders([
                'Authorization' => 'Token ' . config('app.netbox_api_key'),
                'Content-Type' => 'application/json',
            ])
                ->withOptions([
                    'verify' => false,
                    'http_errors' => true,
                ])
                ->patch($url, $options);

The code that the dd command produces is:

"{"device":"58","name":"Bridge-Aggregation1","enabled":true,"label":"Some text","speed":"20000000","mac_address":"AABBCCDDEEFF"}" // app/Models/Netbox.php:468
"https://xxx.xxx.int/api/dcim/interfaces/1725/" // app/Models/Netbox.php:468

**Reply by candlerb** [reply, marked as answer]
> The code that the dd command produces is:

The things that strike me immediately are:
1. you are sending `"device": "58"` instead of `"device": 58`.
2. when patching an interface, I would normally send `"id": <interface-id>` rather than attempting to identify it by a device+name combination

Therefore, although the error message complains about the mac_address, it might actually be something else that's wrong with the request and the error message is a red herring.

You definitely want to compare what your working swagger client sends, with what you're sending from PHP (using the tcpdump command I gave before)

**Reply by NordloSydost-Pajje** [reply]
Hi @candlerb,

Thanks for your suggestion. I did as you said and found the problem. I made an error in my logic for checking if there was an MAC-address included in my file or not. It was always true which in turn made the PHP send "unknown" as a MAC-address.

I also followed your suggestion on skipping device id as i already had the interface id in the URL in the API request.

Thanks again for your time!

BR
Per-Åke
